### PR TITLE
📖 Refactor the BootstrapConfig contract

### PR DESCRIPTION
**Pull request by fabriziopandini**
This PR refactor the BootstrapConfig contract. Most notably

- Everything related to the BootstrapConfig contract is now in one page, or at least in section of the book (before it was spread all around)
- There is a table that gives an overview of all the contract rules that applies to the BootstrapConfig
- Duplicated informations about the BootstrapConfig contract have been removed from other pages

/area documentation

### CODE REVIEW

File: docs/book/src/developer/providers/contracts/bootstrap-config.md [hunk 1]
```diff
@@ -1,107 +1,438 @@
...
 
-## Data Types
+The [Machine's controller](../../core/controllers/machine.md) will be responsible to coordinate operations of the BootstrapConfig,
+and the interaction between the Machine's controller and the BootstrapConfig resource is based on the contract
+rules defined in this page.
 
-### Bootstrap API resource
-A bootstrap provider must define an API type for bootstrap resources. The type:
+Once contract rules are satisfied by an BootstrapConfig implementation, other implementation details
+could be addressed according to the specific needs (Cluster API in not prescriptive).
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply docs/book/src/developer/providers/contracts/bootstrap-config.md:15]: fixed also in other pages

File: docs/book/src/developer/providers/contracts/bootstrap-config.md [hunk 2]
```diff
@@ -1,107 +1,438 @@
...
+1. Have the label `cluster.x-k8s.io/cluster-name` set to the name of the cluster
+1. Have a controller owner reference to the API resource
+1. Have a single key, `value`, containing the bootstrap data
+
+Note: because the `dataSecretName` is part of `status`, this value must be deterministically recreatable from the data in the
+`Cluster`, `Machine`, and/or bootstrap resource. If the name is randomly generated, it is not always possible to move
+the resource and its associated secret from one management cluster to another.
+
+When the Secret is created its name MUST surface in the `status.dataSecretName` field of the BootstrapConfig resource;
+the Machine controller will bubble up this info in Machine's `spec.providerID` when [BootstrapConfig initialization completed].
```
[the remark this answers is not in the document]
Comment (fabriziopandini) [reply docs/book/src/developer/providers/contracts/bootstrap-config.md:214]: fixed also in other pages

**Review by chrischdi — COMMENTED** [review]
/lgtm

**Review by vincepri — APPROVED** [review]
/approve

**Comment by sbueringer** [comment]
/lgtm
in general, follow-up for nits here: https://github.com/kubernetes-sigs/cluster-api/pull/11274

**Pull request merged**